### PR TITLE
Support TypeScript 4.1

### PR DESCRIPTION
**Pull request by clydin**
This change updates the framework to support TypeScript 4.1.

### CODE REVIEW

**Comment by JoostK** [comment]
@googlebot I consent.

File: packages/compiler-cli/ngcc/src/host/commonjs_umd_utils.ts [hunk 1]
```diff
@@ -78,7 +78,7 @@ export function findNamespaceOfIdentifier(id: ts.Identifier): ts.Identifier|null
 export function findRequireCallReference(id: ts.Identifier, checker: ts.TypeChecker): RequireCall|
     null {
   const symbol = checker.getSymbolAtLocation(id) || null;
-  const declaration = symbol && symbol.valueDeclaration;
+  const declaration = symbol?.valueDeclaration ?? symbol?.declarations?.[0];
```
Comment (JoostK) [inline packages/compiler-cli/ngcc/src/host/commonjs_umd_utils.ts:81]: This change seems to have been due to microsoft/TypeScript#39770, where the variable declaration is now treated as an alias declaration which no longer has a value declaration.

**Review by alan-agius4 — COMMENTED** [review]
Can we add an integration tests like https://github.com/angular/angular/tree/master/integration/typings_test_ts40

**Review by gkalpak — APPROVED** [review]
LGTM for `dev-infra`, `docs-infra`, `fw-upgrade`.

From a quick look, I would expect the [`paths` without `baseUrl`][1] breaking change to affect `ngcc`, because we iirc we assume `baseUrl` will be present. For example, `ngcc` [seems to ignore `paths`][2] if there is no `baseUrl`.

Reviewed-for: dev-infra, docs-infra, fw-upgrade

[1]: https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/#paths-without-baseurl
[2]: https://github.com/angular/angular/blob/3e1e5a15ba1d5a08daf9179083e3394df10299bb/packages/compiler-cli/ngcc/src/path_mappings.ts#L22

File: packages/compiler-cli/test/transformers/program_spec.ts [hunk 2]
```diff
@@ -297,15 +297,27 @@ describe('ng program', () => {
...
         'verify that program structure is reused within tsc in order to speed up incremental compilation',
         () => {
+          afterEach(resetTempProgramHandlerForTest);
+
+          function captureStructureReuse(compile: () => void): StructureIsReused|null {
+            let structureReuse: StructureIsReused|null = null;
+            setTempProgramHandlerForTest(program => {
+              structureReuse = tsStructureIsReused(program);
+            });
+            compile();
```
[the remark this answers is not in the document]
Comment (JoostK) [reply packages/compiler-cli/test/transformers/program_spec.ts:307]: That would work, but we'd have to be sure that it is exception safe so the `afterEach` is still nice to have IMO (or wrap it in try-finally)

File: aio/package.json [hunk 3]
```diff
@@ -164,7 +164,7 @@
     "tree-kill": "^1.1.0",
     "ts-node": "^8.4.1",
     "tslint": "~6.1.0",
-    "typescript": "~4.0.2",
+    "typescript": "~4.1.2",
```
[the remark this answers is not in the document]
Comment (clydin) [reply aio/package.json:167]: good point.  Reverted.

File: integration/typings_test_ts41/include-all.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import * as common from '@angular/common';
+import * as commonHttp from '@angular/common/http';
+import * as commonTesting from '@angular/common/testing';
+import * as commonHttpTesting from '@angular/common/testing';
+import * as compiler from '@angular/compiler';
+import * as compilerTesting from '@angular/compiler/testing';
+import * as core from '@angular/core';
+import * as coreTesting from '@angular/core/testing';
+import * as elements from '@angular/elements';
+import * as forms from '@angular/forms';
```
[the remark this answers is not in the document]
Comment (clydin) [reply integration/typings_test_ts41/include-all.ts:23]: good question.  Wasn't there in the 4.0 tests but we can add it.

File: integration/typings_test_ts41/include-all.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import * as animations from '@angular/animations';
+import * as animationsBrowser from '@angular/animations/browser';
+import * as animationsBrowserTesting from '@angular/animations/browser/testing';
+import * as common from '@angular/common';
+import * as commonHttp from '@angular/common/http';
+import * as commonTesting from '@angular/common/testing';
+import * as commonHttpTesting from '@angular/common/testing';
+import * as compiler from '@angular/compiler';
+import * as compilerTesting from '@angular/compiler/testing';
+import * as core from '@angular/core';
```
Comment (gkalpak) [inline integration/typings_test_ts41/include-all.ts:20]: What about `@angular/core/global` 😇:

```suggestion
import * as core from '@angular/core';
import * as coreGlobal from '@angular/core/global';
```

(And also export below.)

File: integration/typings_test_ts41/include-all.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import * as coreTesting from '@angular/core/testing';
+import * as elements from '@angular/elements';
+import * as forms from '@angular/forms';
+import * as localize from '@angular/localize';
```
Comment (gkalpak) [inline integration/typings_test_ts41/include-all.ts:24]: Don't we also need to import from `@angular.localize/init`?

```suggestion
import * as localizeInit from '@angular/localize/init';
```

(And also export below.)

File: integration/typings_test_ts41/include-all.ts (hunk 5 above)
Comment (clydin) [reply integration/typings_test_ts41/include-all.ts:20]: Thinking about this some more, I think for the purposes of this PR (enabling TS 4.1 use) we should keep the 4.0 and 4.1 tests consistent.
And then a follow-up PR can fix both (potentially in a more long-term maintainable fashion).  Components, for instance, has a more generic approach to the concept: https://github.com/angular/components/tree/bb2bb9c3065ec1a7a845de072305893a7378383a/integration/ts-compat

**Comment by jpike88** [comment]
Just checking... I got the error `Internal Error: The structure of the program changed during codegen` when attempting to use typescript 4.1.2 with angular cli 11.0.2. Is this expected, and covered in the PR's fixes? I can't seem to find anyone mentioning that kind of error, and not sure what to make of it.

**Comment by JoostK** [comment]
> Just checking... I got the error `Internal Error: The structure of the program changed during codegen` when attempting to use typescript 4.1.2 with angular cli 11.0.2. Is this expected, and covered in the PR's fixes? I can't seem to find anyone mentioning that kind of error, and not sure what to make of it.

Yes that is expected for ViewEngine compilations, but not Ivy. And yes that is covered in this PR.

**Review by JiaLiPassion — APPROVED** [review]
LGTM
Reviewed-for: zone-js

**Review by alxhub — APPROVED** [review]

**Review by IgorMinar — APPROVED** [review]
LGTM! Thank you!

Reviewed-for: global-approvers

**Pull request closed without merging**